File: 이정선/10월3주차/행렬 테두리 회전하기.cpp
#include <bits/stdc++.h>

using namespace std;

// x1, y1, x2, y2
// (2, 2), (2, 4), (5, 2), (5, 4) 이렇게가 꼭짓점 -> (x1, y1), (x1, y2), (x2, y1), (x2, y2)
// 시계방향 회전
// 꼭짓점 기준으로 이동 계산

vector<int> solution(int rows, int columns, vector<vector<int>> queries) {
    vector<int> answer;
    
    vector<vector<int> > map(rows, vector<int>(columns, 0));
    for(int i = 0; i < rows; i++){
        for(int j = 0; j < columns; j++){
            map[i][j] = (i * columns + j + 1);
        }
    }
        
    int index = 0;
    
    while(index < queries.size()){
        int x1 = queries[index][0];
        int y1 = queries[index][1];
        int x2 = queries[index][2];
        int y2 = queries[index][3];
                
        int minValue = rows * columns;
        
        vector<vector<int> > mapCopy = map;
        
        x1 -= 1;
        y1 -= 1;
        x2 -= 1;
        y2 -= 1;
        
        for(int i = x1; i <= x2; i++){
            for(int j = y1; j <= y2; j++){
                // (x1, y1) 기준
                // x를 한 칸 위로 이동 
                if(i != x2 && j == y1) {
                    mapCopy[i][j] = map[i + 1][j];
                    minValue = min(minValue, mapCopy[i][j]);
                    continue;
                }
                // (x1, y2) 기준
                // y를 한 칸 오른쪽으로 이동 
                if(i == x1 && j != y1) {
                    mapCopy[i][j] = map[i][j - 1];
                    minValue = min(minValue, map[i][j - 1]);
                    continue;
                }
                // (x2, y1) 기준
                // y를 한 칸 왼쪽으로 이동 
                if(i == x2 && j != y2) {
                    mapCopy[i][j] = map[i][j + 1];
                    minValue = min(minValue, map[i][j + 1]);
                    continue;
                }
                // (x2, y2) 기준
                // x를 한 칸 아래쪽으로 이동 
                if(i != x1 && j == y2) {
                    mapCopy[i][j] = map[i - 1][j];
                    minValue = min(minValue, map[i - 1][j]);
                    continue;
                }
                
            }
        }
        map = mapCopy;
        answer.push_back(minValue);
        index++;
    }

    return answer;
}

File: 송하연/10월 3주차/행렬테두리회전하기.java
import java.util.*;

class Solution {
    

    
    public int[] solution(int rows, int columns, int[][] queries) {
        
        int[] answer = new int[queries.length];       
        int[][] matrix = new int[rows][columns];
        
        //행렬 값 채우기
        int cnt=1;
        for(int i=0; i<rows; i++){
            for(int j=0; j<columns; j++){
                matrix[i][j]=cnt;
                cnt++;
            }
        }
        
        //로테이션->배치->최솟값 구하기
        for(int j=0; j<queries.length; j++) {
            int[] query = queries[j];
            List<Integer> list = rotate(query[0],query[1],query[2],query[3],matrix);              
            int[][] rotatedMatrix = place(query[0],query[1],query[2],query[3],matrix,list);  
            answer[j] = Collections.min(list);
        }
        
 
        return answer;
    }
    
    List<Integer> rotate(int x1, int y1, int x2, int y2, int[][] matrix){
        List<Integer> list = new ArrayList<>();

        for (int i=y1; i<y2; i++) list.add(matrix[x1-1][i-1]);    
        for (int i=x1; i<x2; i++) list.add(matrix[i-1][y2-1]);    
        for (int i=y2; i>y1; i--) list.add(matrix[x2-1][i-1]);     
        for (int i=x2; i>x1; i--) list.add(matrix[i-1][y1-1]);  
        
        int lastNum = list.remove(list.size()-1);
        list.add(0, lastNum); 
        
        return list;     
    }
    
    int[][] place(int x1, int y1, int x2, int y2, int[][] matrix, List<Integer> list){
        int idx =0;   
        
        for (int i=y1; i<y2; i++) matrix[x1-1][i-1] = list.get(idx++);
        for (int i=x1; i<x2; i++) matrix[i-1][y2-1] = list.get(idx++);
        for (int i=y2; i>y1; i--) matrix[x2-1][i-1] = list.get(idx++);
        for (int i=x2; i>x1; i--) matrix[i-1][y1-1] = list.get(idx++);
        
        return matrix;
    }  
   
}
